### PR TITLE
governance proposal

### PR DESCRIPTION
**Pull request by mziolekda**

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# https://help.github.com/articles/about-codeowners/
+# Note that accounts or teams mentioned must have WRITE access to the repository.
+
+# Wildcard rule protecting the root of the repository
+*            @mziolek-da @daravep @filmackay
```
Comment (mziolekda) [inline CODEOWNERS:5]: It is meant as an opt in. so volunteers are welcome

File: GOVERNANCE.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+### Seeding
+One of the root code owners creates a directory intended to contain the library and adjusts the CODEOWNERS file to list the designated maintainers.
+
+### Development
+The library is developed in the `daml-common` repository. This process must include
+* providing code with unit tests
+* providing documentation in form of a README file as well as documenting the code forming the library interface
+* adjusting the release definition to contain the new artifacts
+* adjusting the continuous integration to build, test and release the new library
+* adjusting the 3rd party libraru whitelist specifying the required dependencies
```
Comment (rgugliel-da) [inline GOVERNANCE.md]:
```suggestion
* adjusting the 3rd party library whitelist specifying the required dependencies
```

File: GOVERNANCE.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+## Quality
+
+The daml-common repository sets a very high bar on the quality of the libraries included. It should not become a dump of half-baked ideas or where circular dependencies are resolved for lack of a better place. All libraries must
+* use standardized build tooling
+* use blessed 3rd party dependencies
+* be sufficiently tested
+* be sufficiently documented
+* follow standardized coding format
+* adhere to standardized compiler and wart remover flags
```
Comment (daravep) [inline GOVERNANCE.md:65]: IMHO we need to add a "release section!" and we need to add a build system and release method.

**Review by daravep — APPROVED** [review]
LGTM, thanks! I think this is a great start.

File: GOVERNANCE.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+* migration strategy for the existing DA projects
+* a list of third party libraries that the new library will bring as a dependency
+* build system requirements
+* proposed list of the library maintainers that will become code owners on the corresponding directory
+
+### Discussion
+The inclusion of a new library is discussed on the GitHub issue. All DA engineering is invited to take part. In order to be considered conclusive a discussion must garner the attention of at least 5 engineers.
+
+### Acceptance
+Special role in the process is given to the root code owners of this repository. They assume the role of the arbiters of the discussion surrounding inclusion of the new libraries. It does not mean that they have the authocratic power over the contents of the repository. It rather means that they are the gate keepers of the process that can asses that a sufficient consensus has been reached in the github issue discussion. Think of their role as akin to that shared by the speaker of the house of commons in the UK.
```
Comment (da-tanabe) [inline GOVERNANCE.md]: ~authocratic~ autocratic

**Review by da-tanabe — COMMENTED** [review]
How are consumers expected to use this repo?

It might be nice to see something about how artifacts are to be consumed, and what the release process over this repo is (who's "allowed" to cut and/or request a release?)

EDIT: Ah I see @daravep beat me to asking the questions as well!

File: GOVERNANCE.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+`daml-common` is a set of base libraries developed in-house at DA that are meant for wide adoption across all DA products. By selecting a core set of "blessed" libraries we encourage a consistent experience for the development team as well as for the clients using our software. The types of libaries that we envisage will find their home in this repository include but are not limited to:
+
+* logging, metrics, tracing
+* resource management
+* database access abstractions
+* configuration and CLI tooling
+* TLS and access token utilities
+* self-service error codes
+* security evidencing
+* functional utilities such as non-empty monad
```
Comment (andreaslochbihler-da) [inline GOVERNANCE.md]: It's not really a monad.
```suggestion
* functional utilities such as non-empty wrapper
```

File: GOVERNANCE.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+## Process
+
+### Submission
+Inclusion of a new library in the `daml-common` repository must be proposed through a GitHub issue. A valid submission should contain
+* the name of the library to be included
+* a github reference, in case the library already exists in one of the DA repositories
+* justification of why this library must be included and why existing libraries do serve the purpose at the moment
```
Comment (andreaslochbihler-da) [inline GOVERNANCE.md]: I'm missing a negation here.
```suggestion
* justification of why this library must be included and why existing libraries do not serve the purpose at the moment
```

File: GOVERNANCE.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+* configuration and CLI tooling
+* TLS and access token utilities
+* self-service error codes
+* security evidencing
+* functional utilities such as non-empty monad
+
+## Process
+
+### Submission
+Inclusion of a new library in the `daml-common` repository must be proposed through a GitHub issue. A valid submission should contain
```
Comment (andreaslochbihler-da) [inline GOVERNANCE.md:19]: You also want to have the integration into the build and release system of daml-common.

The governance doc makes the impression that all languages are welcome. This means that if we get the first submission for say Haskell, we need to extend the build system to be able to manage and release Haskell stuff. Is this intended? Or do we restrict this repo to say Scala with a single, well-defined build and release process?

File: GOVERNANCE.md (hunk 3 above)
Comment (isegall-da) [reply GOVERNANCE.md:65]: We also need a process for contributions to existing libraries. This file currently focuses on introducing new ones.

File: GOVERNANCE.md (hunk 6 above)
Comment (mziolekda) [reply GOVERNANCE.md]: :+1:

File: GOVERNANCE.md (hunk 5 above)
Comment (mziolekda) [reply GOVERNANCE.md]: :+1:

File: GOVERNANCE.md (hunk 4 above)
Comment (mziolekda) [reply GOVERNANCE.md]: :+1:

File: GOVERNANCE.md (hunk 7 above)
Comment (mziolekda) [reply GOVERNANCE.md:19]: I think we restrict it to scala to start with

File: GOVERNANCE.md (hunk 3 above)
Comment (mziolekda) [reply GOVERNANCE.md:65]: I would add the build system in a separate PR. This PR is just about the governance doc. Although writing something up about the governance of the release and maintenance process is what should go in here. Let me write something up

Comment (mziolekda) [reply GOVERNANCE.md:65]: done

**Pull request merged**